pkg/machines/components/diskRemove.jsx

```diff
@@ -37,8 +38,28 @@ const onDetachDisk = (dispatch, vm, target, onAddErrorNotification) => {
 };
 
 const RemoveDiskAction = ({ dispatch, vm, target, idPrefixRow, onAddErrorNotification }) => {
+    const getRemoveButton = (disabled) => {
+        return (
+            <button id={`${idPrefixRow}-detach`}
+                    disabled={disabled}
+                    className="btn btn-default btn-control-ct fa fa-minus"
+                    onClick={onDetachDisk(dispatch, vm, target, onAddErrorNotification)} />
+        );
+    };
+
+    if (vm.state == 'shut off' || vm.state == 'running')
+        return getRemoveButton(false);
+
     return (
-        <button id={`${idPrefixRow}-detach`} className="btn btn-default btn-control-ct fa fa-minus" onClick={onDetachDisk(dispatch, vm, target, onAddErrorNotification)} />
+        <OverlayTrigger
+            overlay={
+                <Tooltip id="tip-inforec">
+                    {cockpit.format(_("Detaching disks is not supported for $0 VMs"), vm.state)}
```

Comment:
The "not supported" is a bit misleading, it is supported but not available/possible with the current VM state.  I'm wondering whether we want to have such check in cockpit, there is nothing like that in virt-manager.  The reason is that even if the VM is running it can still have broken OS or no OS at all and in that case you will not be able to live-detach the disk as well.

Reply:
@Antique, thanks for the feedback, I was thinking as well if this check has it's place in cockpit. Since libvirt knows that detaching disks with the live flag from paused VMs will fail anyway, maybe it should return error instead of timeing out.

So, either here or in libvirt the check should exist I believe, since  that pausing will timeout for sure. And for running Vms, where as you said it might fail if OS is broken, we 'll get the fail notification anyway, after the timeout expires. It's just a best effort to prevent the user from trying stupid things.

Do you have some better suggestion? 

PS: I would like to avoid to introduce the logic of persistent XMLs vs live XMLs here as we have for example in the NIC configuration.

That all said I think the check should be somewhere for not allowing the operation, if the place is not cockpit but libvirt, I am ok with that as well.

Reply:
It's not that simple, yes we have a timeout in libvirt after witch libvirt returns from the API with success but the device is not actually removed.  The complex part is that the VM OS can just postpone the operation and may remove the device once the timeout expired so it will eventually be removed.  Even with the paused VM it's not that simple, once the VM is resumed the disk will be eventually removed as well.

Probably the only sensible solution in UI is to model it in a way that it will not guarantee the removal but simply "asks" the disk to be removed and it may happen or not.  If the removal is successful libvirt will generate event VIR_DOMAIN_EVENT_ID_DEVICE_REMOVED.  Even if the API returns immediately there is no guarantee that the disk will be actually removed as the VM OS can simply ignore that request.

So the check probably cannot live anywhere and the UI has to inform the user about the conditions and that it is only best-effort operation.

Reply:
@Antique I don't know how we can expose such information to the user nicely, expressing that the X disk might get detached, but it might not.

I would like to keep it as deterministic as possible, even with the cost of removing some functionality, aka detaching disks from paused VMs, and hoping that the operation will take efffect after resume.
 Of course for the current patch, the tool tip wording needs fixing.

That said, I agree that I should enhance the detach feature, so that for running VMs it will inform the user that the operation is a best effort thing.

But I still think that this patch should stay, I tend to want to disallow detaching disks from paused VMs since it can give us only headaches.
@martinpitt do you have any feedback on this?

Reply:
Not a strong one, perhaps @garrett can chime in here. My gut feeling is that it's nice and cheap for the cases where it's definitively not working (paused). Then the tooltip should say "not possible" instead of "not supported". For the running case, the button should maybe say "Request detaching"?

Reply:
Or show some notifications about the progress of the operation. (cockpit-composer does this actually)

First we can show a info notification that disk detach has started which will expire and disappear after X seconds. In the mean time we should show an in progress' icon, while waiting for the DEVICE_REMOVED signal. If we don't receive the signal the API fail with timeout, we just show the timeout fail notification and remove the 'in progress' icon.

But in any case, let's not introduce this logic for paused VMs. I am happy to implement it (either this or change the button to 'Request detaching' for the running.

Reply:
The API will never fail with timeout error, in this case it will always return success.

As a note, is it OK to remove the 'in progress' icon once the API returns and no DEVICE_REMOVED signal arrived? The signal may arrive in a few seconds after that happen but it can also never arrive or in several minutes/hours.